### PR TITLE
Allow building with OpenSSL 1.1

### PR DESCRIPTION
**Pull request by JanMarvin**
This PR should enable building RStudio with OpenSSL <= 1.1 (at least it works with OpenSSL 1.1). I could not yet check this actual PR but a similar version located in my Arch Linux PKBGUILD [1]. To be completely able to build RStudio with OpenSSL 1.1 you are still required to patch around the missing `SSL_R_SHORT_READ`. Locally I use a modified version of #1063. 
Possibly this requires further tweaking and I feel a bit uneasy messing around with ssl. For now it is just what works in my non public environment. 

[1] https://github.com/JanMarvin/rstudio-archlinuxpkg/blob/master/rstudio-server/openssl11.diff

### CODE REVIEW

File: src/cpp/core/system/PosixCrypto.cpp [hunk 1]
```diff
@@ -257,6 +257,11 @@ core::Error rsaInit()
    const int KEY_SIZE = 1024;
    const int ENTROPY_BYTES = 4096;
 
+   BIGNUM *bn = BN_new();
```
Comment (jjallaire) [inline src/cpp/core/system/PosixCrypto.cpp]: It looks like this is never freed if the first branch of the `#if OPENSSL_VERSION_NUMBER` branch is taken.

File: src/cpp/core/system/PosixCrypto.cpp [hunk 2]
```diff
@@ -271,14 +276,29 @@ core::Error rsaInit()
...
+      bn_n = s_pRSA->n;
+      bn_e = s_pRSA->e;
+   #else
+      s_pRSA = RSA_new();
+      int rc = ::RSA_generate_key_ex(s_pRSA, KEY_SIZE, bn, NULL);
+      if (rc != 1) {
+        return lastCryptoError(ERROR_LOCATION);
+        RSA_free(s_pRSA);
+      }
+      BN_clear_free(bn);
```
Comment (jjallaire) [inline src/cpp/core/system/PosixCrypto.cpp]: This never gets called if `rc != 1` above.

**Comment by jjallaire** [comment]
@jmcphers or @kevinushey Could you review this as well?

@JanMarvin We need a contributor agreement from you before merging (see https://github.com/rstudio/rstudio/blob/master/CONTRIBUTING.md)

**Comment by JanMarvin** [comment]
I was wondering if I need to free `bn_n`and `bn_e` as well for the `#if OPENSSL_VERSION_NUMBER` part?

@jjallaire you should have a contributor agreement from me. Dating from around 2014?

**Comment by jmcphers** [comment]
`bn_n` and `bn_e` look to just be pointers into the static struct. I don't think they need to be freed explicitly. The docs for `get0_key` suggest as much: https://www.openssl.org/docs/man1.1.0/crypto/RSA_get0_key.html

One other nit: you could just move the allocation of `bn` into the else block since it's only used there (otherwise there's a correct but pointless alloc and free of the struct when running against older openssl).

**Comment by kevinushey** [comment]
Looks good to me as well (with the same comments re: alloc + free of `bn` into `#else` block)

**Comment by JanMarvin** [comment]
Thanks, updated.

File: src/cpp/core/system/PosixCrypto.cpp [hunk 3]
```diff
@@ -271,14 +274,32 @@ core::Error rsaInit()
...
+   #else
+      BIGNUM *bn = BN_new();
+      BN_set_word(bn, RSA_F4);
+ 
+      s_pRSA = RSA_new();
+      int rc = ::RSA_generate_key_ex(s_pRSA, KEY_SIZE, bn, NULL);
+      BN_clear_free(bn);
+      if (rc != 1) {
+        return lastCryptoError(ERROR_LOCATION);
+        RSA_free(s_pRSA);
```
Comment (kevinushey) [inline src/cpp/core/system/PosixCrypto.cpp]: This code won't be reached, given the prior `return` -- should this be moved up a line?

Comment (JanMarvin) [inline src/cpp/core/system/PosixCrypto.cpp]: Thanks, moved it.

**Comment by jjallaire** [comment]
Yes, we do have a contributor agreement already. Thanks!

**Pull request merged**